File: 2025-12-December-GFG-POTD/December 28 - Minimum time to fulfil all orders/Solution.java
class Solution {
     public int minTime(int[] ranks, int n) {
        int minRank = Integer.MAX_VALUE;
        for (int r : ranks) {
            minRank = Math.min(minRank, r);
        }

        int low = 0;
        int high = minRank * n * (n + 1) / 2;
        int ans = high;

        while (low <= high) {
            int mid = low + (high - low) / 2;

            if (canMake(ranks, n, mid)) {
                ans = mid;
                high = mid - 1; // try smaller time
            } else {
                low = mid + 1;  // need more time
            }
        }
        return ans;
    }

    private boolean canMake(int[] ranks, int n, int time) {
        int donuts = 0;

        for (int r : ranks) {
            int k = 0;
            int currTime = 0;

            while (true) {
                currTime += r * (k + 1);
                if (currTime > time) break;
                k++;
            }

            donuts += k;
            if (donuts >= n) return true;
        }
        return false;
    }
}

File: 2025-12-December-GFG-POTD/December 28 - Minimum time to fulfil all orders/Solution.py
class Solution:
    def minTime(self, ranks, n):
        minRank = min(ranks)

        low = 0
        high = minRank * n * (n + 1) // 2
        ans = high

        while low <= high:
            mid = low + (high - low) // 2

            if self.canMake(ranks, n, mid):
                ans = mid
                high = mid - 1
            else:
                low = mid + 1

        return ans

    def canMake(self, ranks, n, time):
        donuts = 0

        for r in ranks:
            k = 0
            currTime = 0

            while True:
                currTime += r * (k + 1)
                if currTime > time:
                    break
                k += 1

            donuts += k
            if donuts >= n:
                return True

        return False
